File: 034/find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var search = function (nums, target) {
    let [low, high] = [0, nums.length - 1];
    let mid;
    while (low <= high) {
        mid = (low + high) >> 1;
        if (nums[mid] < target) {
            low = mid + 1;
        } else if (nums[mid] > target) {
            high = mid - 1;
        } else return mid;
    }
    return -1;
};

var searchRange = function (nums, target) {
    let index = search(nums, target);
    if (index == -1)
        return [-1, -1];

    let first = index;
    while (true) {
        let firstCandidate = search(nums.slice(0, first), target);
        if (firstCandidate == -1) break;
        first = firstCandidate;
    }

    let last = index;
    while (true) {
        let lastCandidate = search(nums.slice(last+1), target);
        if (lastCandidate == -1) break;
        last += lastCandidate + 1;
    }

    return [first, last];
};


let nums = [5, 7, 7, 8, 8, 10], target = 8;
console.log(searchRange(nums, target));

File: 034/find-first-and-last-position-of-element-in-sorted-array.go
package main

import "fmt"

func searchRange(nums []int, target int) []int {
	pickOne := search(nums, target)
	if pickOne == -1 {
		return []int{-1, -1}
	}

	first := pickOne
	for {
		firstCandidate := search(nums[:first], target)
		if firstCandidate == -1 {
			break
		}
		first = firstCandidate
	}

	last := pickOne
	for {
		lastCandidate := search(nums[last+1:], target)
		if lastCandidate == -1 {
			break
		}
		last = lastCandidate
	}

	return []int{first, last}
}

// binart search in sorted array
func search(nums []int, target int) int {
	low, high := 0, len(nums)-1
	var mid int
	for low <= high {
		mid = (low + high) >> 1
		switch {
		case nums[mid] < target:
			low = mid + 1
		case nums[mid] > target:
			high = mid - 1
		default:
			return mid
		}
	}
	return -1
}

func main() {
	nums := []int{5, 7, 7, 8, 8, 10}
	target := 8
	fmt.Println(searchRange(nums, target))
}
